### ISSUE TITLE
ThyraTpetraAdapters_TpetraThyraWrappersUnitTests_MPI_4 failing in CI build pushed on 4/20/2017

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#1258]
**Description:**

The test `ThyraTpetraAdapters_TpetraThyraWrappersUnitTests_MPI_4` started failing in the first CI iteration pushed in the morning of 4/20/2017 shown here:

* http://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=2848932&filtercount=3&showfilters=1&field1=groupname&compare1=61&value1=Continuous&field2=buildstarttime&compare2=84&value2=now&filtercombine=and

The new commits pulled this CI iteration are shown here:

* http://testing.sandia.gov/cdash/viewNotes.php?buildid=2849008##note1

The commit range for the new commits pulled are 7ecd901923fc9dd00b5215d7252165ad2b2912b7...e6afc15358ddddd78000e7036391920558bf2ff1

The test output for the broken test is shown here:

* http://testing.sandia.gov/cdash/testDetails.php?test=38144658&build=2849008

shows:

```
  D) Testing the MultiVectorBase interface of mv ...
   
   *** Entering Thyra::MultiVectorTester<double>::check(mv,...) ...
   
   Testing a MultiVectorBase object mv described as:
    Thyra::TpetraMultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >{rangeDim=16,domainDim=4}
   
   Checking the LinearOpBase interface of mv ...
    
    *** Entering LinearOpTester<double,double>::check(op,...) ...
    
    describe op: Thyra::TpetraMultiVector<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >{rangeDim=16,domainDim=4}
    
    Checking the domain and range spaces ... passed!
    
    this->check_linear_properties()==true:Checking the linear properties of the forward linear operator ... passed!
    
    (this->check_linear_properties()&&this->check_adjoint())==true: Checking the linear properties of the adjoint operator ... passed!
    
    this->check_adjoint()==true: Checking the agreement of the adjoint and forward operators ... 
    op.opSupported(CONJTRANS) = true == true : passed
    
    Checking that the adjoint agrees with the non-adjoint operator as:
    
      <0.5*op'*v2,v1> == <v2,0.5*op*v1>
       \________/            \_______/
           v4                   v3
    
             <v4,v1>  == <v2,v3>
    
    Random vector tests = 1
     
     v1 = randomize(-1,+1); ...
     
     v2 = randomize(-1,+1); ...
     
     v3 = 0.5*op*v1 ...
     
     v4 = 0.5*op'*v2 ...
     
     Check: rel_err(<v4,v1>, <v2,v3>)
            = rel_err(-2.09678, -0.524194) = 0.75
              <= adjoint_error_tol() = 2.22045e-12 : FAILED
    
    this->check_for_symmetry()==false: Skipping check of symmetry ...
    
    Oh no, at least one of the tests performed with this LinearOpBase object failed (see above failures)!
    
    *** Leaving LinearOpTester<double,double>::check(...)
   
   Oh no, at least one of the tests performed with this MultiVectorBase object failed (see above failures)!
   
   *** Leaving MultiVectorTester<double>::check(mv,...)
```

This suggests that the adjoint operation for Tpetra::MultiVector is broken.

**CC:** @trilinos/framework, @trilinos/tpetra, @trilinos/thyra

**Comment by bartlettroscoe** [comment]
Since this is failing test blocking anyone else from pushing with the checkin-test-sems.sh script for packages upstream from Thyra, we should resolve this soon.  This can be done with one of the following:

1. Fix the change in Tpetra that is breaking this test (usually best option)
2. Backout commits that are breaking this test and then fix the issues offline and push again (best if issue can't fixed quickly and broken feature is important)
3. Temporarily disable this test but fix it soon (if this is not a critical feature)

This test suggests the the Tpetra::MultiVector adjoint application is broken.  The fact that this is the only test that broke in Trilinos suggests that either there is no algorithmic code in Trilinos that actually uses adjoints with multi-vectors or this code is not tested in Trilinos.  Therefore, it is not clear how critical this feature is and if it warrants backing out the changes if it can't be fixed fairly quickly (i.e. before the night builds fire off).

**Comment by mhoemmen** [comment]
I started working on this with @rppawlo yesterday.  I'm able to replicate a strange issue in Tpetra itself, and I need a little time to debug it.  If you need to revert, just let me know.

**Comment by mhoemmen** [comment]
@bartlettroscoe There's also a quick fix I can make without losing the value of the commit.  Would you like to me to try that?

**Comment by bartlettroscoe** [comment]
> There's also a quick fix I can make without losing the value of the commit. Would you like to me to try that?

@mhoemmen, if that does not take too much time, then that would be great.   Since the commits were rebased instead of merged in, it would be a little harder to revert them all one-at-a-time.

NOTE: the remote pull/test/push process always creates an explicit merge commit for all of the commits you are pushing so it makes them easy to back out if needed.

**Comment by mhoemmen** [comment]
@bartlettroscoe Thanks!  I'll implement the quick fix.  The change in question was the result of a team room session with @crtrott and @vbrunini so it was a little more hasty than my usual process.  I conflated some useful harmless things with something that _looked_ harmless but wasn't.  Ah well ;-)

**Comment by mhoemmen** [comment]
I pushed a quick fix.  It restores the original behavior of Tpetra::MultiVector::dot, without removing the good stuff @crtrott et al. helped me add.  I'll work on figuring this out, since the optimization improved performance by about a quarter on GPUs (by avoiding >=1 device View allocation per dot product call).

**Comment by mhoemmen** [comment]
@bartlettroscoe Does Thyra implement randomize on vectors itself?  It looks like Thyra's Tpetra wrappers are not calling Tpetra::MultiVector::randomize at all.  Nothing wrong with that in itself, but I'm curious whether it's related to this issue.

**Comment by bartlettroscoe** [comment]
From @mhoemmen:

> Does Thyra implement randomize on vectors itself? It looks like Thyra's Tpetra wrappers are not calling Tpetra::MultiVector::randomize at all. Nothing wrong with that in itself, but I'm curious whether it's related to this issue. It could be that something isn't getting sync'd when it should be. This could also relate to Thyra GPU correctness so it's important that I dig into it a little bit.

There is no randomize() virtual function on any of the main Thyra interfaces.  Therefore, I believe that this Thyra testing code uses an RTOp for Randomize and that is indeed what this looks like in the implementation (see line 125):

* https://trilinos.org/docs/dev/packages/thyra/doc/html/Thyra__VectorSpaceTester__def_8hpp_source.html#l00074

The randomize() function that uses a dumb RTOp is not sufficient for any type of fancy numerical algorithm (e.g. that might generate linearly independent columns, for example).  It is just good enough for some simple testing where you just need to fill a bunch of values between upper and lower with no particular strong statement about how these values are distributed.

But I think these are valid multi-vectors so if this simple adjoint test fails, it might indicate that there may be some type of problem.

BTW, it looks like this test got fixed as of this CI iteration last Friday:

* http://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=2850976

and you can see the test is newly passing in:

* http://testing.sandia.gov/cdash/viewTest.php?onlypassed&buildid=2850981

The only new commit that could have affected this was:

```
e1b2e8b:  Tpetra::MultiVector::dot: Temp fix for #1258
Author: Mark Hoemmen <mhoemme@sandia.gov>
Date:   Fri Apr 21 11:27:51 2017 -0600

M	packages/tpetra/core/src/Tpetra_MultiVector_def.hpp
```

Thanks for that!

If this needs a more substantial fix, that can be tracked in a separate issue.

Thanks @mhoemmen!

**Comment by mhoemmen** [comment]
I am fixing the underlying performance issue that caused #1258.  See #1389.
